src/Microsoft.IdentityModel.Tokens/CryptoProviderFactory.cs

```diff
@@ -383,22 +384,34 @@ private SignatureProvider CreateSignatureProvider(SecurityKey key, string algori
             if (typeofSignatureProvider == null)
                 throw LogHelper.LogExceptionMessage(new NotSupportedException(LogHelper.FormatInvariant(LogMessages.IDX10621, typeof(SymmetricSignatureProvider), typeof(SecurityKey), typeof(AsymmetricSecurityKey), typeof(SymmetricSecurityKey), key.GetType())));
 
+            if (!IsSupportedAlgorithm(algorithm, key))
+                throw LogHelper.LogExceptionMessage(new NotSupportedException(LogHelper.FormatInvariant(LogMessages.IDX10634, algorithm, key)));
+
             if (CacheSignatureProviders)
             {
                 if (CryptoProviderCache.TryGetSignatureProvider(key, algorithm, typeofSignatureProvider, willCreateSignatures, out signatureProvider))
                     return signatureProvider;
-            }
-
-            if (!IsSupportedAlgorithm(algorithm, key))
-                throw LogHelper.LogExceptionMessage(new NotSupportedException(LogHelper.FormatInvariant(LogMessages.IDX10634, algorithm, key)));
+                
+                lock (_cacheLock)
+                {
+                    if (CryptoProviderCache.TryGetSignatureProvider(key, algorithm, typeofSignatureProvider, willCreateSignatures, out signatureProvider))
+                        return signatureProvider;
 
-            if (createAsymmetric)
-                signatureProvider = new AsymmetricSignatureProvider(key, algorithm, willCreateSignatures);
-            else
-                signatureProvider = new SymmetricSignatureProvider(key, algorithm, willCreateSignatures);
+                    if (createAsymmetric)
+                        signatureProvider = new AsymmetricSignatureProvider(key, algorithm, willCreateSignatures);
+                    else
+                        signatureProvider = new SymmetricSignatureProvider(key, algorithm, willCreateSignatures);
 
-            if (CacheSignatureProviders)
-                CryptoProviderCache.TryAdd(signatureProvider);
+                    if (CacheSignatureProviders)
```

Comment:
This check is redundant.

Reply:
Thanks for noticing this! Just fixed this.